### ISSUE TITLE
set up accrual accounting

### ISSUE THREAD

**Issue by whit537** [gratipay/inside.gratipay.com#308]
Right now we account for our operations on a cash basis with a [spreadsheet](http://inside.gratipay.com/appendices/finances). Sooner or later we should implement a proper accrual system. I'm thinkin' [Xero](https://www.xero.com/) (cf. https://github.com/gratipay/gratipay.com/issues/1039).

**Comment by techtonik** [comment]
What is accruel? Can it be simplified? We need a system to do accounting, right?

**Comment by whit537** [comment]
There are two kinds of accounting: cash (simple) and accrual (robust). We've been using cash accounting so far. Accrual accounting is more powerful but more complex.

https://en.wikipedia.org/wiki/Basis_of_accounting

**Comment by techtonik** [comment]
I really bad at reading that accounting terminology. Found this https://www.khanacademy.org/economics-finance-domain/core-finance/accounting-and-financial-stateme/cash-accrual-accounting/v/accrual-basis-of-accounting

**Comment by techtonik** [comment]
Looks like a problem with cash accounting is that cash doesn't have tags. So you can't track what money is for what, what are debts, what are payments in advance, for which demands. It looks like you just mix all stuff in and can only track income and expenses at certain intervals. I see why it is mostly useless. =)

**Comment by whit537** [comment]
I've gone back and dusted off our Xero account:

![screen shot 2015-09-22 at 10 22 37 am](https://cloud.githubusercontent.com/assets/134455/10021021/e3cd67c2-6113-11e5-8c92-846533effe31.png)

**Comment by whit537** [comment]
I've linked our PNC account and our PayPal account.

**Comment by whit537** [comment]
Alright, this is getting interesting. I've been reconciling our PNC account with Xero, and I've been handling the influx of money from PayPal as a transfer from a PayPal account to a bank account. But when I get to reconciling our PayPal account in Xero, what am I going to do with the influxes of money from Gratipay?

One option might be to add Gratipay as a "bank account" so that I can handle it as a transfer:

![screen shot 2015-09-22 at 10 45 03 am](https://cloud.githubusercontent.com/assets/134455/10021660/0954fc8c-6117-11e5-9e0e-c4b2b01139db.png)

In that case, Gratipay would export a feed that Xero could consume, and we'd reconcile transactions within Xero just like with any other bank (or PayPal) account. Depending on how someone was using Gratipay, they'd have income and expense transactions, eventually including payroll expense. We've been talking about Gratipay vis-a-vis banks (with reference to PayPal) over at https://github.com/gratipay/gratipay.com/issues/1383#issuecomment-141482842.

Xero also has a "[payment services](https://help.xero.com/us/PaymentServices)" feature under their invoice template settings, and it includes a "custom URL" option:

![screen shot 2015-09-22 at 10 48 33 am](https://cloud.githubusercontent.com/assets/134455/10021756/8cd11a78-6117-11e5-878d-b3646d81ef32.png)

This assumes that the business is invoicing customers up front and giving them options to pay. Gratipay implements self-billed, pay-what-you-want invoicing, so the model doesn't quite fit. They do have a guide on [handling payment processor receipts](http://developer.xero.com/documentation/api-guides/handling-payment-processor-receipts-in-xero/), but even then, Gratipay isn't a simple payment processor, because of a) payroll, and b) receiving and giving in the same cycle. If The Enterprise receives $1,000 on Gratipay, gives $100, and distributes $800 in payroll, then their PayPal account will only see $100, but they need to report $1,000 in income and $900 in expenses, not merely $100 in income.

**Comment by whit537** [comment]
Soooo ... Gratipay is "[not a bank](https://medium.com/gratipay-blog/charging-in-arrears-18cacf779bee)," but it plays one ~~on TV~~ in your company's books?

**Comment by whit537** [comment]
I've reconciled the most recent info that Xero was able to bring over from PNC, but I'd like to go back and enter and reconcile all of our past history that we've maintained on [our finances spreadsheet](https://docs.google.com/spreadsheets/d/1p3DpF9ZLEsViBx0685FwJaYN1vVScKVUgXHb2zyqXZg/pub?output=html&widget=true).

**Comment by whit537** [comment]
As of https://github.com/gratipay/gratipay.com/pull/3786#issuecomment-142299405 I'm thinking of this as the over-arching ticket for whatever accounting features we need to add to gratipay.com. We need a proper accounting system, and accounting for our escrow is a part of that. Therefore, the next step would seem to be setting up Xero to include both our operating and escrow accounts, and then the conversation with the CPA can be about making sure our books are set up sensibly. I'm going to put a hold on https://github.com/gratipay/gratipay.com/pull/3786 for now, and focus on Xero.

**Comment by whit537** [comment]
> The total balance in escrow is shown on a company's balance sheet at the end of every period, either as a liability or in a section between liabilities and equity. The separate bank account that the escrow funds are held in is shown as a current asset outside of the company's operating account. In this way, the escrow funds are both an asset and a liability in the company and are a "wash." However, the accounts cannot be netted against each other because they must be visible as both asset and liability according to GAAP.

http://www.ehow.com/info_11384335_accounting-treatment-escrow-account.html

That and http://www.tdi.texas.gov/title/documents/BASICS.pdf are the best info I've found so far on setting up books for an escrow.

**Comment by whit537** [comment]
> Pooled escrow accounts may require individual client ledgers.

http://www.ehow.com/info_12119449_accounting-journal-entry-amounts-held-escrow.html

**Comment by whit537** [comment]
To: Peter Fleming, CPA

> Are we still on for tomorrow at 1:30pm?

**Comment by whit537** [comment]
Xero is going to have both our operating and escrow accounts in it. It'll take some work to bring all of our operating history into Xero, but that's not the interesting stuff. More important for tomorrow is the escrow accounting, so I'm going to focus on that first. I'm going to import our escrow history into Xero.

**Comment by whit537** [comment]
Alright, what are all the accounts we've ever had escrow in?

 - Samurai
 - Stripe
 - Balanced
 - Braintree
 - New Alliance FCU
 - Ally
 - Citizens
 - PayPal
 - Coinbase
 - Transferwise

**Comment by whit537** [comment]
The purpose right now is to account for everything at a high level. We don't need per-user ledgers yet. Right now we just want to get all of our top-level escrow activity into one place: Xero.

**Comment by whit537** [comment]
I want to see escrow accounts on our balance sheet:

![screen shot 2015-09-22 at 12 40 05 pm](https://cloud.githubusercontent.com/assets/134455/10024865/2389d770-6127-11e5-894a-a94044236066.png)

**Comment by whit537** [comment]
I'm looking at our PayPal transaction history. I don't think we want individuals ~users in Xero. I think we want lump sums in Xero, with individual ~users accounted for in gratipay.com itself.

**Comment by whit537** [comment]
I've started a download of our PayPal history.

**Comment by whit537** [comment]
Okay, I have PayPal history, but only back two years. :-/

**Comment by whit537** [comment]
PayPal's web interface goes back three years, which is long enough. Our first PayPal payout was on Jan 20, just after [payday 33](https://github.com/gratipay/gratipay.com/issues/490) (not mentioned there nor in 34).

**Comment by whit537** [comment]
There's actually a little bit of noise in the PayPal prior to Jan 20, which is reflect on our finances spreadsheet. When did we entangle escrow from operations?

**Comment by whit537** [comment]
"[sort out bank accounts](https://github.com/gratipay/gratipay.com/issues/68)"

**Comment by whit537** [comment]
Also helpful for early history:

"[go over payday numbers and make sure it is right](https://github.com/gratipay/gratipay.com/issues/171)"

... and especially ... 

"[correct any misuse of escrowed funds](https://github.com/gratipay/gratipay.com/issues/265)," which this ticket should enable if we don't in fact end up needing to resolve 265 before closing this here.

**Comment by whit537** [comment]
I believe we were mingling escrow and operations for over a year, until https://github.com/gratipay/gratipay.com/issues/68 two years ago.

**Comment by whit537** [comment]
https://github.com/gratipay/gratipay.com/issues/171 links to [an old spreadsheet](https://docs.google.com/spreadsheets/d/1ZjXxKGCbDYub1XMcwO6v1wkrq_qtv-Ll5kkTBNXPpr4/) which seems to line up with a portion of 2012 from [our finances spreadsheet](https://docs.google.com/spreadsheets/d/1p3DpF9ZLEsViBx0685FwJaYN1vVScKVUgXHb2zyqXZg/), except that the beginning balance is given as $106.23, which is not what is computed on our finances spreadsheet.

**Comment by whit537** [comment]
New Alliance history appears to only go back a little over a year.

**Comment by whit537** [comment]
I guess we'll have to reconstruct from Balanced and PayPal, etc.

**Comment by whit537** [comment]
I've kicked off an export of our Balanced data.

**Comment by whit537** [comment]
I'm considering making two separate Xero accounts to represent New Alliance as an operating account, and as an escrow account.

**Comment by whit537** [comment]
I've exported New Alliance back through 7/7/2014. The opening balance at that point is 41,746.78.

**Comment by whit537** [comment]
I've downloaded our entire Coinbase history.

**Comment by whit537** [comment]
History data:

 - [x] Samurai—long gone, will have to reconstruct from Finances
 - [x] Stripe
 - [ ] Balanced—waiting for the export to complete; is that process still working at Balanced?
 - [ ] Braintree
 - [ ] New Alliance FCU—could only get back through 7/7/2014, will have to reconstruct from other sources. We have a balance at 6/21/2012 of $106.23.
 - [x] Ally—we have these in PDF statements and possibly a csv (need to double-check it)
 - [ ] Citizens—I think we have this in PDFs?
 - [ ] PayPal—We have back through 9/24/2013 in a CSV, need to scrape the rest from the web UI
 - [x] Coinbase
 - [ ] Transferwise—only one of these
 - [ ] cash—a couple of these very early

**Comment by whit537** [comment]
On the phone with New Alliance. $10/hr (might take an hour or two?) and a couple days to get transaction history going back that far.

**Comment by whit537** [comment]
Alright, went ahead and pulled the trigger on that. Even if we end up reconstructing from other sources for tomorrow, it'll be good to have that as a check.

**Comment by whit537** [comment]
To: Balanced Support

> Hey gang! I've kicked off an export and am waiting for an email saying it's done. Waiting longer than I remember having to before. Thought I'd check: are exports still working? :-)

**Comment by whit537** [comment]
I have a CSV for Ally, spot-check looks good comparing with statements. However, we only have statements through April, 2015 with a balance of $82,372.56, and it looks like we closed the account in May, 2015 with a final balance (per New Alliance, where the money went) of $82,414.81. A pro-rated interest payment of $42.25 would explain the difference, but I may still call Ally to see if we can pin that down for sure.

**Comment by whit537** [comment]
[Looks like](https://gratipay.freshdesk.com/helpdesk/tickets/2133) we did receive a May statement, but I didn't download it. :-/

**Comment by whit537** [comment]
Called Ally and ordered a resend of our last statement. Either 1-2 days by email, or 7-10 days by snail mail.

**Comment by whit537** [comment]
We only have one month for Citizens, but I think that's all it was: $30,000 in and out, [right quick](https://github.com/gratipay/gratipay.com/issues/3366#issuecomment-113762413). :-)

**Comment by whit537** [comment]
I think we can treat Samurai, Stripe, and Braintree as [simple payment processors](http://developer.xero.com/documentation/api-guides/handling-payment-processor-receipts-in-xero/), with batch settlements/disbursements hitting our New Alliance bank account in Xero.

Since we kept/keep balances at Balanced and PayPal, and use them for both payins and payouts, I guess we should handle those as Current Asset accounts.

Actually, Xero calls for payment processors to have a Current Asset clearing account in our books, so I guess they're all similar after all.

New Alliance, Ally, and Citizens will of course be straight-up bank accounts. Not sure exactly how we'll deal with Transferwise and cash yet.

**Comment by whit537** [comment]
I'm trying to understand the difference between Bank and Current Asset accounts in Xero.

> **Bank accounts**
>
A bank account in Xero has specific behavior to make and receive payments which you can reconcile to the statement lines coming from your actual bank account. You can also specify other accounts to record payments against internally in Xero when you add an account or edit an existing account.

https://help.xero.com/us/Settings_ChartofAccounts$BK_XOintro

**Comment by whit537** [comment]
I set up "Escrow - New Alliance FCU" as a bank account, and did a manual import of the CSV. It worked pretty well!

**Comment by whit537** [comment]
I had to manually configure Xero integration with PayPal for the vendors@gratipay.com PayPal account (our operations one), but when I went to configure it for paypal@gratipay.com (escrow), it was already configured! :eyes:

**Comment by whit537** [comment]
I've been reconciling the New Alliance import. I'm down from 201 transactions to 34, and have just hit the transition from "BALANCED" to "BRAINTREE FUNDING". I had Balanced in there as a bank account. Is that right? Is that what I should do for Braintree?

**Comment by whit537** [comment]
I'm fine to leave Balanced as a pseudo-bank account because we want to move money forth *and back*. A question for Peter, though, I suppose.

**Comment by whit537** [comment]
Okay! Last transaction left is Transferwise. How are we going to account for that?

**Comment by whit537** [comment]
I updated the conversation balance (opening balance) for NAFCU, and it matches our current balance!

![screen shot 2015-09-22 at 6 41 54 pm](https://cloud.githubusercontent.com/assets/134455/10033805/abffda14-6159-11e5-9dc4-67e1b400a44c.png)

**Comment by whit537** [comment]
Alright, so that's the thing to do:

 - use May 31, 2012 as the conversion date
 - for each account, set the conversion balance to the earliest record we have, even if it's not that far back
 - enter the transactions, such as we have them
 - reconcile and balance everything
 - go back and add past transactions once we have them

**Comment by whit537** [comment]
Alright, we're [on our own](https://www.youtube.com/watch?v=VL0NSk8_50s).

> Thanks for checking in.  I apologize for the delayed response.  I have been embroiled in a project recently that has kept me out of the office a lot.  I will need to reschedule with you for this afternoon – very sorry for the last minute change.  I have 2 emergencies today that needed to be resolved.
>
I am open all day next Thursday – 10/1.  Will that work?

-----------

> Okay, thanks for letting me know. I've got deadlines next Thursday and the Thursday after that to deal with this escrow I'm sitting on. I'm going to get through it as best I can and then perhaps we can sit down for a post-mortem in late October. Let's touch base again in a couple weeks.
>
Best of luck with your projects! :-)

**Comment by whit537** [comment]
I love where this is headed with Xero.

![screen shot 2015-09-23 at 9 11 58 am](https://cloud.githubusercontent.com/assets/134455/10045783/3430bc36-61d3-11e5-83d0-5e37436b1879.png)

**Comment by whit537** [comment]
Alright, I've entered PayPal data back to 7/7/2014 (going back further had problems because there were no matching records in history for New Alliance for transfers to and fro). I ended up writing a script to process the PayPal export into a format that worked to import into Xero. I have four unreconciled transactions, MassPay refunds. How are fees handled in that case?

![screen shot 2015-09-23 at 10 02 39 am](https://cloud.githubusercontent.com/assets/134455/10047054/4084c8d6-61da-11e5-8bc5-542e80f2b953.png)

**Comment by whit537** [comment]
Reconcile data back through 7/7/2014:

 - [x] Samurai—no data that old
 - [x] Stripe—no data that old
 - [ ] Balanced
 - [ ] Braintree
 - [ ] New Alliance FCU
 - [ ] Ally
 - [ ] Citizens
 - [ ] PayPal
 - [ ] Coinbase
 - [ ] Transferwise

**Comment by whit537** [comment]
I made a "~user" contact to use with the "299 - Escrow" liability account, for example, with this Transferwise transaction.

**Comment by mattbk** [comment]
!m @whit537!  That's a lot of old bank stuff to wade through!

**Comment by whit537** [comment]
@kaguillera and I are at Blue Canary working on this. I'm going to give @kaguillera access to the logs repo so he can convert the Ally and Citizens statements from PDF to CSV. I did pick up the New Alliance statements going back to January 2012. They're on paper, and will need to be entered as CSV before we can bring that data into Xero.

**Comment by whit537** [comment]
If we can get reconciled and balanced back to 7/7/2014 for both Escrow and Operations, then we can tackle importing data from 6/1/2012 through 7/6/2014 in a separate step.

**Comment by whit537** [comment]
We need to think about how fees and escrow are reported from Gratipay over to Xero.

**Comment by whit537** [comment]
The end goal of this ticket is an accurate Balance Sheet. We're getting there! :-)

![screen shot 2015-09-23 at 1 59 19 pm](https://cloud.githubusercontent.com/assets/134455/10053703/87d10fd0-61fb-11e5-8d84-cd757cf86f94.png)

**Comment by whit537** [comment]
To: GitHub Support

> Any chance we could have early access to the PDF comment attachments feature for the `gratipay` org? :-)

**Comment by whit537** [comment]
- @whit537 to clean up `Escrow - Coinbase` and `Escrow - PayPal`
 - @whit537 to generate a CSV for `Escrow - Balanced`—maybe by hacking their JSON API?
 - @whit537 to add PDF statements for `PNC` to the logs repo (we were only able to get a CSV export back to 2015-07)
 - @kaguillera to start on producing CSVs for operations accounts:
  - `Ally`
  - `Citizens`
  - `PNC`
  - `New Alliance`

**Comment by whit537** [comment]
I'm sorting out a $1.16 discrepancy in Coinbase. There are indeed two $1.16 transactions in New Alliance on Oct 28, and also in Coinbase. What's up with Xero?

**Comment by whit537** [comment]
But there are *three* $1.16 transactions in Xero.

**Comment by whit537** [comment]
Hey @kaguillera, we have pretty much all of operations in [the finances spreadsheet](https://docs.google.com/spreadsheets/d/1p3DpF9ZLEsViBx0685FwJaYN1vVScKVUgXHb2zyqXZg/edit). The ones we don't have are the past few months from PNC.

**Comment by whit537** [comment]
Alright! Coinbase is reconciled! I was accidentally creating a third $1.16 transaction during reconciliation.

**Comment by whit537** [comment]
> [I made](https://github.com/gratipay/inside.gratipay.com/issues/308#issuecomment-142612877) a "~user" contact to use with the "299 - Escrow" liability account, for example, with this Transferwise transaction.

I moved away from the "~user" contact, instead using a contact for each payin or payout method. Currently:

 - Braintree
 - PayPal MassPay
 - PayPal Instant Transfer
 - Coinbase Payout
 - Coinbase Payin
 - Transferwise

That makes it easier to track sources of escrow liability.

**Comment by whit537** [comment]
P.S. Xero appears to have brought in the latest escrow PayPal transactions automatically, so that's working.

**Comment by whit537** [comment]
Let's figure out how to handle these PayPal MassPay refunds ...

**Comment by whit537** [comment]
From: GitHub Support
Subject: Re: PDF attachments?

> We're being methodical about a gradual rollout for this and there's currently no way to accelerate that for a specific repository or user/org. I'll definitely pass your interest on to the team working on this feature though.

**Comment by whit537** [comment]
I've dug up one of the four MassPays in Xero that includes a failed payout.

![screen shot 2015-09-24 at 9 07 32 am](https://cloud.githubusercontent.com/assets/134455/10074106/d3cb0210-629b-11e5-932f-d9114817b29d.png)

**Comment by whit537** [comment]
The [detailed log](https://github.com/gratipay/logs/blob/130db070ca20b00e7949a188e2df827ccc525020/masspay/2015-08-13.report.paypal.csv) shows that there is no fee assessed for the failed payout.

**Comment by whit537** [comment]
It turns out we haven't been using the actual fee when posting back to Gratipay, we've been using our guess. I've reticketed using the actual fee as https://github.com/gratipay/gratipay.com/issues/3796. For this ticket, this could be a source of some slight discrepancies when we start to balance escrow assets and liabilities. Let's keep our :eyes: out!

**Comment by whit537** [comment]
For now I've reconciled the four MassPay refunds against `299 - Escrow Liability` account, with "PayPal MassPay" as the contact (same as successful MassPay payouts).

**Comment by whit537** [comment]
Okay! The last escrow asset to reconcile is Balanced! :o:

**Comment by whit537** [comment]
Making progress on Balanced. I was able to download all of our statements (invoices) through their JSON API, and now I am preparing a CSV to import into Xero and reconcile.

One note: Balanced took our fee out of our New Alliance account, independently of any transfers between Balanced and New Alliance (we were responsible for shuffling money back and forth). By contrast, Coinbase, PayPal, Braintree, and Stripe all take the fee out from the balance in their system before settling out to us.

**Comment by whit537** [comment]
That means we don't need to include the fee in `balanced.csv`, since it's already present in New Alliance.

**Comment by whit537** [comment]
It also means that the Balanced invoices don't include information about when we manually moved money back and forth with New Alliance.

**Comment by whit537** [comment]
Do we need that in order to reconcile?

**Comment by whit537** [comment]
And: have we accounted for Braintree's fee? Oh, that's right ... they [haven't started charging us fees yet](https://www.braintreepayments.com/blog/ignition-first-50k-on-us/).

**Comment by whit537** [comment]
> Do we need that in order to reconcile?

Looks like yes:

![screen shot 2015-09-24 at 11 15 59 am](https://cloud.githubusercontent.com/assets/134455/10077307/bf388e50-62ad-11e5-96c0-d4b552fde848.png)

**Comment by whit537** [comment]
I've mixed the transfers to and from New Alliance into `balanced.csv` and have imported and reconciled that statement. Still something wrong ...

![screen shot 2015-09-24 at 12 12 20 pm](https://cloud.githubusercontent.com/assets/134455/10079001/a2ffdbdc-62b5-11e5-9796-9671ee35c8c5.png)

**Comment by whit537** [comment]
P.S. That's with a conversion balance set, as calculated from the full JSON records.

**Comment by whit537** [comment]
Eep! Now `Escrow - New Alliance` and `Escrow - PayPal` are out as well! A clue! :-)

![screen shot 2015-09-24 at 12 33 11 pm](https://cloud.githubusercontent.com/assets/134455/10079515/81fcb768-62b8-11e5-80ec-9d0d950d88de.png)

**Comment by whit537** [comment]
The problem seems to be with conversion balances. They have this "comparative balances" thing that you can only *add* from the conversion balances UI ([old blog post](https://www.xero.com/blog/2008/07/delete-comparative-balances/) has a broken video about deleting). I'm poking around in the "historical adjustments" account and it looks like there's a bogus entry in there that's probably screwing us up ...

**Comment by whit537** [comment]
Looks like I need to make a manual journal entry to reverse a bogus trial balance adjustment.

**Comment by whit537** [comment]
Using the "conversion balances" UI results in two journal entries, both on the conversion date: one to reverse the previous conversion balance journal entry, and a second for the new conversion balances. Comparative balance entries manifest as "trial balance adjustment" journal entries, and we're stuck with on on December 31, 2014. I'm going to try reversing that ...

![screen shot 2015-09-24 at 12 57 26 pm](https://cloud.githubusercontent.com/assets/134455/10080173/10070e48-62bc-11e5-968e-70ac4e6e5a3d.png)

![screen shot 2015-09-24 at 12 59 51 pm](https://cloud.githubusercontent.com/assets/134455/10080186/2be71888-62bc-11e5-88f8-66f1d227aa90.png)

**Comment by whit537** [comment]
The journal entry screen is not allowing me to specify an asset account. O.O

**Comment by whit537** [comment]
> Although you can't post a manual journal directly to a bank account in Xero, there are three ways to achieve the same effect.

https://help.xero.com/us/Q_JournalBankAcc

**Comment by whit537** [comment]
Blech. :disappointed:

**Comment by whit537** [comment]
Okay, I [was able](https://community.xero.com/business/discussion/3584174) to reverse the bogus trial balance adjustment. Now we're back to just having Balanced out:

![screen shot 2015-09-24 at 1 40 24 pm](https://cloud.githubusercontent.com/assets/134455/10081219/dd5688ec-62c1-11e5-948e-427ba430f36b.png)

**Comment by whit537** [comment]
Interesting. Looks like an edge case with our July 7, 2014 cutoff date:

![screen shot 2015-09-24 at 1 43 14 pm](https://cloud.githubusercontent.com/assets/134455/10081302/45efbf22-62c2-11e5-8903-ff9c78d54fab.png)

**Comment by whit537** [comment]
Okay! I imported a second statement with just that transaction, and now Xero and Balanced are supposedly reconciled. However, the balance is off compared to the Balanced dashboard. We at least need to update the conversion balance to include that latest transaction ...

**Comment by whit537** [comment]
Alright! Now we're out by ~$1,000 ...

![screen shot 2015-09-24 at 1 55 11 pm](https://cloud.githubusercontent.com/assets/134455/10081649/e7f3ddc0-62c3-11e5-8fb9-6c0a2760cd3f.png)
![screen shot 2015-09-24 at 1 54 47 pm](https://cloud.githubusercontent.com/assets/134455/10081648/e7f2c642-62c3-11e5-8c70-06e08ba25e1b.png)

**Comment by whit537** [comment]
$770.04, to be precise.

**Comment by whit537** [comment]
I think the next step is to get detailed transaction history out of Balanced.

**Comment by whit537** [comment]
The problem exists in `balanced.csv`, because the ending balance in there is indeed 27,632.58. We don't have all of the transactions we need in `balanced.csv`. Why not?

**Comment by whit537** [comment]
Maybe chargebacks and disputes?

**Comment by whit537** [comment]
Looks like we have 117,077 transactions at Balanced, and we can get 100 at a time through the JSON API. I've started downloading them. Maybe 20 minutes to get them all?

**Comment by whit537** [comment]
I've confirmed that we *are* getting failed as well as successful transactions (which is good).

**Comment by whit537** [comment]
I think requests are taking longer than a second each, so it could be more like 30 or 40 minutes.

**Comment by whit537** [comment]
Lunchtime! :-)

**Comment by whit537** [comment]
662 / 1171 (56%)

**Comment by whit537** [comment]
72%

**Comment by whit537** [comment]
Here's the keys in the various transaction dicts so far:

```python
set([u'card_holds',
     u'credits',
     u'debits',
     u'links',
     u'meta',
     u'refunds',
     u'reversals'])
```

**Comment by whit537** [comment]
I'm writing a simple script to compute the balance.

**Comment by whit537** [comment]
Hrm. Should've downloaded by `created_at` *ascending*.

**Comment by whit537** [comment]
Alright, first attempt at computing our Balance balance from a list of all transactions yields $28,358.98. 

| source | balance | ∆ |
| Balanced dashboard | 28,402.62 | 0 |
| transactions/*.json | 28,358.98 | -43.64 |
| Xero | 27,632.58 | -770.14 |

**Comment by whit537** [comment]
Filtering for `status == 'succeeded'`:

| source | balance ($) | ∆ ($) |
|:---|:---|---:|
| transactions/*.json | 28,389.43 | -13.19 |

**Comment by whit537** [comment]
Well, that's maddening. :-)

**Comment by whit537** [comment]
Did the download drop something? Is (gasp) the Balanced dashboard wrong?

**Comment by whit537** [comment]
Here's my script: https://gist.github.com/whit537/432657942af5ebdc7a8a.

**Comment by whit537** [comment]
To: Balanced Support

> I'm working on setting up a proper accounting system for Gratipay, and I'm coming up with a $13.19 discrepancy between the balance reported in our Balanced dashboard, and the balance I'm able to compute from a download of all our transactions (I grabbed them using the JSON API since the CSV export to email isn't working for me; see my other ticket).
>
Here's the script I'm using to compute the balance from the transactions JSON:
>
https://gist.github.com/whit537/432657942af5ebdc7a8a
>
The result of that script is $28,389.43, and our dashboard shows $28,402.62.
>
Any idea what I'm missing? Thanks! :-)

**Comment by whit537** [comment]
Searching for 13.19 in the Balanced dashboard returns 328 transactions.

**Comment by whit537** [comment]
Do we have them all in JSON?

**Comment by whit537** [comment]
I count 155 debits with amount 13.19 in the JSON. The dashboard results included holds. How about without holds?

**Comment by whit537** [comment]
I count 155 successful debits in the dashboard as well.

**Comment by whit537** [comment]
I've checked the transaction ids of the JSON and the dashboard, and they line up. It seems that all 13.19 transactions in the dashboard are accounted for in the JSON.

**Comment by whit537** [comment]
13.19 is a fairly common charge amount (it's an upcharge of $12.50). It *seems* like one of these credit card transactions would be implicated in the discrepancy, rather than a bank deposit (or reversal).

**Comment by whit537** [comment]
Since there's more in the dashboard than in the JSON, that indicates that the dashboard thinks a charge succeeded that the JSON doesn't have or thinks didn't succeed. Or the JSON thinks a refund succeeded that the dashboard doesn't.

**Comment by whit537** [comment]
> I've checked the transaction ids of the JSON and the dashboard [...]

I loaded up all the transactions I wanted in the dashboard and then copied the HTML using the DOM inspector. I pasted that into a local file and munged it with ack/grep and vim. It seemed that a bug prevented me from in fact filtering out `failed` transactions in the dashboard listing, so I dealt with those in  the final comparison (for which I used `diff`): there were eight.

**Comment by whit537** [comment]
We could try the download again, possibly varying the page size.

**Comment by whit537** [comment]
/me kicks it off with a page size of 99.

**Comment by whit537** [comment]
We've still got the more significant discrepancy between Balanced (whether transaction-based or dashboard-based) and Xero (which is invoice- and transfer-based).

**Comment by whit537** [comment]
What's up with that?

**Comment by whit537** [comment]
Got the last statement from Ally. Indeed, we had a $42.25 interest payment.

**Comment by whit537** [comment]
I downloaded the transactions again in pages of 99 transactions, and the balance is coming out the same: $28,389.43.

**Comment by whit537** [comment]
Alright, so let's let the $13.19 go for now. What about the $770.14?

**Comment by whit537** [comment]
Refund fees?

**Comment by whit537** [comment]
My next steps here are to collate the transactions in ascending date order, inserting a "type" key, and probably storing them in one file. Then I'd see about recomputing the invoices alongside the invoice-based CSV generator and looking for discrepancies.

**Comment by whit537** [comment]
Huzzah! :dancer: 

[Gratipay, LLC - Balance Sheet.pdf](https://github.com/gratipay/inside.gratipay.com/files/2703/Gratipay.LLC.-.Balance.Sheet.pdf)

**Comment by whit537** [comment]
Boo. :-(

No inline niceness!

**Comment by whit537** [comment]
Gah, now PayPal is out as well. :-(

![screen shot 2015-09-28 at 10 30 40 am](https://cloud.githubusercontent.com/assets/134455/10138343/fd168abc-65cb-11e5-9b94-c65d83da4dff.png)

![screen shot 2015-09-28 at 10 33 03 am](https://cloud.githubusercontent.com/assets/134455/10138409/52877bd2-65cc-11e5-826d-d73f5c555db5.png)

**Comment by whit537** [comment]
There were two transactions in PayPal that weren't brought over to Xero yet via the automatic feed. I turned off the feed and manually exported and imported another statement, and now we're reconciled.

**Comment by whit537** [comment]
Alright, I've caught up PNC and PayPal.

@kaguillera No more PDF conversion is necessary. :-)

**Comment by whit537** [comment]
![screen shot 2015-09-28 at 12 30 34 pm](https://cloud.githubusercontent.com/assets/134455/10141842/c1b694b0-65dc-11e5-8328-c173f9b5bf5f.png)

**Comment by whit537** [comment]
Inputing historical data reticketed as #349.

**Comment by whit537** [comment]
Now to finish reconciling Balanced and then balancing escrow assets and liability.

**Comment by whit537** [comment]
I signed us up for Xero's $30/mo "Standard" plan.

https://www.xero.com/pricing/

**Comment by whit537** [comment]
We'll need to come up with some policies, published on Inside Gratipay, for recording financial transactions and reconciling with our bank accounts. I'm not going to worry about that for this ticket. I've reticketed having an account review our books as #350, let's pick up with that there.

**Comment by whit537** [comment]
So I'm trying to reproduce the statements Balanced gave us to track down this $770.04 (ignoring the $13.19 for now).

**Comment by whit537** [comment]
I find exactly one invoice `period` of `[null, null]`. It has chargeback fees, and they're not in the `fees` substructure that I believe I was looking in to generate transactions from invoices. Let's see ...

**Comment by whit537** [comment]
But we weren't actually looking at the fee amounts, we were looking at the escrow-affecting amounts, credits and debits and reversals and refunds.

**Comment by whit537** [comment]
When we got hit with chargebacks, did Balanced take those out of our escrow directly? Or did they debit our bank account like with other fees?

**Comment by msherry** [comment]
@whit537 Chargebacks are invoiced along with fees, not taken from escrow.

**Comment by whit537** [comment]
A wild @msherry appears! :dancer:

**Comment by whit537** [comment]
@msherry Any idea why summing over an export of all transactions.json is out $13.19 from our balance shown in the dashboard?

https://github.com/gratipay/inside.gratipay.com/issues/308#issuecomment-143031121

**Comment by whit537** [comment]
Alright, so chargebacks are invoiced with fees ... but what about the refunds associated with chargebacks? Those presumably come out of escrow. Do they hit the `invoices.fees` structure?

**Comment by msherry** [comment]
Any funds sent back to your users come from escrow. The operating principle to remember is that money in escrow is earmarked for your customers/merchants, not for operating expenses. By the same token, money sent to your customers (refunds, credits, etc.) will not come from your operating capital, and is not a fee.

**Comment by whit537** [comment]
@msherry Okay, makes sense. The two discrepancies I'm trying to reconcile are:

 1. $770.04 difference between our balance in the Balanced dashboard, and our newly set up books in Xero.
 2. $13.19 difference between our balance in the Balanced dashboard, and a sum over an export of all transactions.json ([my script](https://gist.github.com/whit537/432657942af5ebdc7a8a), assuming json files in `transactions/`).

I'm working on (1). Are you able to confirm my result for (2)?

**Comment by whit537** [comment]
Actually, (1) isn't a discrepancy between the Balanced dashboard and Xero, it's a discrepancy between the Balanced dashboard, and a sum over an export of all invoices. Here's [my script](https://gist.github.com/whit537/ebce0eabdae9d8affb9f) for that. It operates on `invoices-N.json` and `transfers.json` files in the same directory. The transfers are transactions for the marketplace owner account.

**Comment by whit537** [comment]
@msherry Basically, I'm looking for a reliable way to generate a CSV of statements from Balanced that I can import into Xero and reconcile with the rest of our accounts. Right now it's a lie in this screenshot that `Escrow - Balanced` is reconciled, because the balance doesn't match our balance in the Balanced dashboard.

![screen shot 2015-09-28 at 4 27 00 pm](https://cloud.githubusercontent.com/assets/134455/10147728/da31af7c-65fd-11e5-8f63-0fb0a0572c0f.png)

**Comment by whit537** [comment]
Given that the transactions in Xero *are* reconciled between New Alliance and Balanced for the dates in range (July 7, 2014 to present, since that's what we could get easily for New Alliance), it would seem that the discrepancy is either before that date, or in non-New-Alliance-touching transactions.

**Comment by whit537** [comment]
(I entered a conversion balance for Balanced in Xero that was based on the [balance-from-invoices](https://gist.github.com/whit537/ebce0eabdae9d8affb9f) method of working with Balanced statements.)

**Comment by whit537** [comment]
Sure seems like I should be able to compute our current balance from both a) a listing of all statements, and b) a listing of all transactions.

**Comment by whit537** [comment]
Alright. New plan. I'm going to digitize the New Alliance data, so we have something solid to reconcile against. Then I'm going to go through and reconcile month by month starting from the present and working backwards.

**Comment by whit537** [comment]
The conversion balance for New Alliance on 2012-05-31 is $102.45.

**Comment by whit537** [comment]
The [finances spreadsheet](https://docs.google.com/spreadsheets/d/1p3DpF9ZLEsViBx0685FwJaYN1vVScKVUgXHb2zyqXZg/edit#gid=5) has from before June 1, 2012 (our start date) through June 6, 2013 for New Alliance, but then it drops out until picking up with Ally on September 27, 2013 (see https://github.com/gratipay/gratipay.com/issues/68). I've gone through the New Alliance data in the spreadsheet and have compared it with the paper statements I have, and they line up (after I fixed a simple formula error). Therefore, here are the fourteen New Alliance statements we need to digitize:

 - [] June, 2013
 - [] July, 2013
 - [] August, 2013
 - [] September, 2013
 - [] October, 2013
 - [] November, 2013
 - [] December, 2013
 - [] January, 2014
 - [] February, 2014
 - [] March, 2014
 - [] April, 2014
 - [] May, 2014
 - [] June, 2014
 - [] July, 2014

**Comment by whit537** [comment]
> but then it drops out until picking up with Ally on September 27, 2013

That simply means that statements I'm about to digitize will be necessary to backfill operations in Xero as well as escrow.

**Comment by whit537** [comment]
Blam.

**Comment by whit537** [comment]
That wasn't too bad: 2.5 hrs, including supper. :-)

**Comment by whit537** [comment]
Okay! Now let's reconcile some accounts!

Oh! We still need a year of PayPal ...

**Comment by whit537** [comment]
Gah, lost a week of PayPal history since https://github.com/gratipay/inside.gratipay.com/issues/308#issuecomment-142362539. :-(

**Comment by whit537** [comment]
Alright, didn't actually drop any PayPal transaction history in the past week. I've screenscraped the web UI and made a CSV out of it.

**Comment by whit537** [comment]
*Now* let's reconcile some accounts!

**Comment by whit537** [comment]
Actually, let's go to bed. :-)

See you Wednesday, if the creek don't rise!

**Comment by whit537** [comment]
I've got three CSVs for New Alliance:

 1. `newalliance.0.csv`—from the digitized paper statements
 1. `newalliance.1.csv`—from the finances spreadsheet
 1. `newalliance.2.csv`—the web export

From (2) and (3) I've split out two more CSVs, for the operations transactions that were mixed in with the escrow transactions (I've [removed](https://github.com/gratipay/logs/commit/a801124f4d727d0a211c487834f103ba393f4701) those transactions from the CSVs listed above).

I've created a `New Alliance` bank account in Xero, distinct from the `Escrow - New Alliance` bank account. I've reconciled the two CSVs for it, and we are showing a balance of -$1,355.71. This may shift a bit when I get into fees between users and Gratipay (I assigned the FeeFighters fees to operations; they did come out New Alliance in a way somewhat similar to Balanced, but they're so unrelated to the actual deposits from FeeFighters that I thought it easier to deal with them apart from escrow). However, I believe what this is telling us is that we may have spent $1,355.71 out of escrow, which we will need to pay back if we haven't already in some other way (https://github.com/gratipay/gratipay.com/issues/265).

**Comment by whit537** [comment]
Now to reconcile the rest of the `Escrow - New Alliance` transactions, and then the rest of `Escrow - PayPal`.

**Comment by whit537** [comment]
Missed a couple GitHub transactions. Looking at -$1,368.71.

**Comment by whit537** [comment]
Another one. Up to $2,168.71.

**Comment by whit537** [comment]
Alright, after reconciling the rest of `Escrow - New Alliance` and resetting the conversion balance to 0, we are at 74,178.00, which is out by ... $2,168.71. :-)

**Comment by whit537** [comment]
I haven't done a good job of accounting for fees. For example, I'm only recording a 9.41 deposit from Stripe, and I'm losing the 0.59 payment processing fee.

**Comment by whit537** [comment]
Alright, this is ridiculous. :hurtrealbad:

**Comment by whit537** [comment]
Okay! @kaguillera and I got Balanced balanced today. Progress!

[![screen shot 2015-09-30 at 9 05 51 pm](https://cloud.githubusercontent.com/assets/134455/10210612/14c3494c-67b7-11e5-89e6-8caaebbf8c10.png)](https://github.com/gratipay/inside.gratipay.com/files/5628/Gratipay.LLC.-.Balance.Sheet.pdf)

**Comment by whit537** [comment]
I'm not going to get this any further before the [advances refund](https://github.com/gratipay/gratipay.com/issues/3763) tomorrow. I think it could make a lot of sense for @kaguillera and I to double-team these follow-on issues on Wednesdays when we're able to get together. We made a lot more progress together today than I'm able to alone on these accounting issues.

!m @kaguillera
